src/helpers/block_requests.rs

```diff
@@ -0,0 +1,413 @@
+// Copyright (C) 2019-2021 Aleo Systems Inc.
+// This file is part of the snarkOS library.
+
+// The snarkOS library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The snarkOS library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkOS library. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::Environment;
+use snarkvm::dpc::prelude::*;
+
+use std::net::SocketAddr;
+
+///
+/// Handle block requests. Returns the start/end block heights to request and if the ledger requires a fork.
+///
+/// Case 1 - You are ahead of your peer:
+///     - Do nothing
+/// Case 2 - You are behind your peer:
+///     Case 2(a) - `is_fork` is `None`:
+///         - Peer is being malicious or thinks you are ahead. Both are issues,
+///           pick a different peer to sync with.
+///     Case 2(b) - `is_fork` is `Some(false)`:
+///         - Request blocks from your latest state
+///     Case 2(c) - `is_fork` is `Some(true)`:
+///             Case 2(c)(a) - Common ancestor is within `MAXIMUM_FORK_DEPTH`:
+///                  - Revert to common ancestor, and send block requests to sync.
+///             Case 2(c)(b) - Common ancestor is NOT within `MAXIMUM_FORK_DEPTH`:
+///                  Case 2(c)(b)(a) - You can calculate that you are outside of the `MAXIMUM_FORK_DEPTH`:
+///                      - Disconnect from peer.
+///                  Case 2(c)(b)(b) - You don't know if you are within the `MAXIMUM_FORK_DEPTH`:
+///                      - Revert to most common ancestor and send block requests to sync.
+///
+pub async fn handle_block_requests<N: Network, E: Environment>(
+    latest_block_height: u32,
+    latest_cumulative_weight: u128,
+    maximal_peer: Option<SocketAddr>,
+    maximal_peer_is_fork: Option<bool>,
+    maximum_block_height: u32,
+    maximum_cumulative_weight: u128,
+    maximum_common_ancestor: u32,
+    first_deviating_locator: Option<&u32>,
+) -> Option<(u32, u32, bool)> {
+    // Case 1 - Ensure the peer has a heavier canonical chain than this ledger.
+    if latest_cumulative_weight >= maximum_cumulative_weight {
+        return None;
+    }
```

Comment:
this check is already performed in `update_block_requests`; it should be removed from there so as not to duplicate the logic